### ISSUE TITLE
Issues when deploying to Azure Function App

### ISSUE THREAD

**Issue by ferristocrat** [microsoft/PowerShellForGitHub#282]
#### Issue Details
Error when using in an Azure Function App

#### Steps to reproduce the issue
*note that when running the same code locally, I do not get the error. It is only when deployed as a Function App
```powershell
   $githubToken = "blahblahblahtoken"
   $githubOwnerName = "somegithubaccount"
   $githubRepoName = "thisisareponame"

   # Create secure string from token
   $secureString = ($githubToken | ConvertTo-SecureString -AsPlainText -Force)

   # Create credential
   $cred = New-Object System.Management.Automation.PSCredential "username is ignored", $secureString

   # Set GitHub auth for the session (Error occurs here)
   Set-GitHubAuthentication -Credential $cred -SessionOnly

   $secureString = $null # clear this out now that it's no longer needed
   $cred = $null # clear this out now that it's no longer needed

   # Get the GitHub repository
   $githubRepo = Get-GitHubContent -OwnerName $githubOwnerName -RepositoryName $githubRepoName

```

#### Verbose logs showing the problem
ERROR: Cannot process argument transformation on parameter 'Path'. Cannot convert value "" to type "System.IO.FileInfo". Error: "The path is empty. (Parameter 'path')" Exception : Type : System.Management.Automation.ParameterBindingArgumentTransformationException Message : Cannot process argument transformation on parameter 'Path'. Cannot convert value "" to type "System.IO.FileInfo". Error: "The path is empty. (Parameter 'path')" ParameterName : Path ParameterType : System.IO.FileInfo TypeSpecified : string ErrorId : ParameterArgumentTransformationError Line : 498 Offset : 5 CommandInvocation : MyCommand : Write-Log BoundParameters : Comparer : System.OrdinalIgnoreCaseComparer Count : 2 Keys : Length : 7 Length : 5 Values : Length : 1 LongLength : 1 Rank : 1 SyncRoot : Length : 135 IsFixedSize : True Count : 1 Length : 7 SyncRoot : Comparer : System.OrdinalIgnoreCaseComparer Count : 2 Keys : Length : 7 Length : 5 Values : Length : 1 LongLength : 1 Rank : 1 SyncRoot : Length : 135 IsFixedSize : True Count : 1 Length : 7 SyncRoot : Comparer : System.OrdinalIgnoreCaseComparer Count : 2 Keys : Length : 7 Length : 5 Values : Length : 1 LongLength : 1 Rank : 1 SyncRoot : Length : 135 IsFixedSize : True Count : 1 Length : 7 SyncRoot : Comparer : System.OrdinalIgnoreCaseComparer Count : 2 Keys : Length : 7 Length : 5 Values : Length : 1 LongLength : 1 Rank : 1 SyncRoot : Length : 135 IsFixedSize : True Count : 1 Length : 7 SyncRoot : Comparer : System.OrdinalIgnoreCaseComparer Count : 2 Keys : Length : 7 Length : 5 Values : Length : 1 LongLength : 1 Rank : 1 SyncRoot : Length : 135 IsFixedSize : True Count : 1 Length : 7 SyncRoot : Comparer : System.OrdinalIgnoreCaseComparer Count : 2 Keys : Length : 7 Length : 5 Values : Length : 1 LongLength : 1 Rank : 1 SyncRoot : … IsFixedSize : True Count : 1 Length : 7 SyncRoot : Comparer : System.OrdinalIgnoreCaseComparer Count : 2 Keys : … Values : … SyncRoot : … ScriptLineNumber : 498 OffsetInLine : 5 HistoryId : 1 ScriptName : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1 Line : Write-Log -Message "[$($Invocation.MyCommand.Module.Version)] Executing: $($Invocation.MyCommand) $($params -join ' ')" -Level Verbose PositionMessage : At D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1:498 char:5 + Write-Log -Message "[$($Invocation.MyCommand.Module.Version)] Exe … + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ PSScriptRoot : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0 PSCommandPath : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1 InvocationName : Write-Log PipelineLength : 1 PipelinePosition : 1 CommandOrigin : Internal ErrorRecord : Exception : Type : System.Management.Automation.ParentContainsErrorRecordException Message : Cannot process argument transformation on parameter 'Path'. Cannot convert value "" to type "System.IO.FileInfo". Error: "The path is empty. (Parameter 'path')" HResult : -2146233087 CategoryInfo : InvalidData: (:) [Write-Log], ParentContainsErrorRecordException FullyQualifiedErrorId : ParameterArgumentTransformationError,Write-Log InvocationInfo : MyCommand : Write-Log ScriptLineNumber : 498 OffsetInLine : 5 HistoryId : 1 ScriptName : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1 Line : Write-Log -Message "[$($Invocation.MyCommand.Module.Version)] Executing: $($Invocation.MyCommand) $($params -join ' ')" -Level Verbose PositionMessage : At D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1:498 char:5 + Write-Log -Message "[$($Invocation.MyCommand.Module.Version)] Exe … + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ PSScriptRoot : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0 PSCommandPath : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1 CommandOrigin : Internal ScriptStackTrace : at Write-InvocationLog, D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1: line 498 at Set-GitHubAuthentication, D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\GitHubConfiguration.ps1: line 877 at <ScriptBlock>, D:\home\site\wwwroot\Github\run.ps1: line 42 TargetSite : Name : BindParameter DeclaringType : System.Management.Automation.ParameterBinderBase, System.Management.Automation, Version=7.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 MemberType : Method Module : System.Management.Automation.dll StackTrace : at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter, CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags) at System.Management.Automation.CmdletParameterBinderController.BindParameter(CommandParameterInternal argument, MergedCompiledCommandParameter parameter, ParameterBindingFlags flags) at System.Management.Automation.CmdletParameterBinderController.BindParameter(UInt32 parameterSets, CommandParameterInternal argument, MergedCompiledCommandParameter parameter, ParameterBindingFlags flags) at System.Management.Automation.ParameterBinderController.BindUnboundScriptParameterWithDefaultValue(MergedCompiledCommandParameter parameter) at System.Management.Automation.ParameterBinderController.BindUnboundScriptParameters() at System.Management.Automation.CmdletParameterBinderController.BindCommandLineParameters(Collection`1 arguments) at System.Management.Automation.CommandProcessor.BindCommandLineParameters() at System.Management.Automation.CommandProcessor.Prepare(IDictionary psDefaultParameterValues) at System.Management.Automation.CommandProcessorBase.DoPrepare(IDictionary psDefaultParameterValues) at System.Management.Automation.Internal.PipelineProcessor.Start(Boolean incomingStream) at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input) --- End of stack trace from previous location where exception was thrown --- at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input) at System.Management.Automation.PipelineOps.InvokePipeline(Object input, Boolean ignoreInput, CommandParameterInternal[][] pipeElements, CommandBaseAst[] pipeElementAsts, CommandRedirection[][] commandRedirections, FunctionContext funcContext) at System.Management.Automation.Interpreter.ActionCallInstruction`6.Run(InterpretedFrame frame) at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame) Data : System.Collections.ListDictionaryInternal InnerException : Type : System.Management.Automation.ArgumentTransformationMetadataException ErrorRecord : Exception : Type : System.Management.Automation.ParentContainsErrorRecordException Message : Cannot convert value "" to type "System.IO.FileInfo". Error: "The path is empty. (Parameter 'path')" HResult : -2146233087 CategoryInfo : MetadataError: (:) [], ParentContainsErrorRecordException FullyQualifiedErrorId : RuntimeException TargetSite : Name : Transform DeclaringType : System.Management.Automation.ArgumentTypeConverterAttribute, System.Management.Automation, Version=7.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 MemberType : Method Module : System.Management.Automation.dll StackTrace : at System.Management.Automation.ArgumentTypeConverterAttribute.Transform(EngineIntrinsics engineIntrinsics, Object inputData, Boolean bindingParameters, Boolean bindingScriptCmdlet) at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter, CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags) Message : Cannot convert value "" to type "System.IO.FileInfo". Error: "The path is empty. (Parameter 'path')" InnerException : Type : System.Management.Automation.PSInvalidCastException ErrorRecord : Exception : Type : System.Management.Automation.ParentContainsErrorRecordException Message : Cannot convert value "" to type "System.IO.FileInfo". Error: "The path is empty. (Parameter 'path')" HResult : -2146233087 CategoryInfo : InvalidArgument: (:) [], ParentContainsErrorRecordException FullyQualifiedErrorId : InvalidCastConstructorException TargetSite : Name : Convert DeclaringType : System.Management.Automation.LanguagePrimitives+ConvertViaConstructor, System.Management.Automation, Version=7.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 MemberType : Method Module : System.Management.Automation.dll StackTrace : at System.Management.Automation.LanguagePrimitives.ConvertViaConstructor.Convert(Object valueToConvert, Type resultType, Boolean recursion, PSObject originalValueToConvert, IFormatProvider formatProvider, TypeTable backupTable) at System.Management.Automation.LanguagePrimitives.ConversionData`1.Invoke(Object valueToConvert, Type resultType, Boolean recurse, PSObject originalValueToConvert, IFormatProvider formatProvider, TypeTable backupTable) at System.Management.Automation.LanguagePrimitives.ConvertTo(Object valueToConvert, Type resultType, Boolean recursion, IFormatProvider formatProvider, TypeTable backupTypeTable) at System.Management.Automation.ArgumentTypeConverterAttribute.Transform(EngineIntrinsics engineIntrinsics, Object inputData, Boolean bindingParameters, Boolean bindingScriptCmdlet) Message : Cannot convert value "" to type "System.IO.FileInfo". Error: "The path is empty. (Parameter 'path')" InnerException : Type : System.ArgumentException Message : The path is empty. (Parameter 'path') ParamName : path TargetSite : Name : GetFullPath DeclaringType : System.IO.Path MemberType : Method Module : System.Private.CoreLib.dll StackTrace : at System.IO.Path.GetFullPath(String path) at System.IO.FileInfo..ctor(String originalPath, String fullPath, String fileName, Boolean isNormalized) at lambda_method(Closure , Object ) at System.Management.Automation.LanguagePrimitives.ConvertViaConstructor.Convert(Object valueToConvert, Type resultType, Boolean recursion, PSObject originalValueToConvert, IFormatProvider formatProvider, TypeTable backupTable) Source : System.Private.CoreLib HResult : -2147024809 Source : System.Management.Automation HResult : -2147467262 Source : System.Management.Automation HResult : -2146233087 Source : System.Management.Automation HResult : -2146233087 CategoryInfo : InvalidData: (:) [Write-Log], ParameterBindingArgumentTransformationException FullyQualifiedErrorId : ParameterArgumentTransformationError,Write-Log InvocationInfo : MyCommand : Write-Log ScriptLineNumber : 498 OffsetInLine : 5 HistoryId : 1 ScriptName : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1 Line : Write-Log -Message "[$($Invocation.MyCommand.Module.Version)] Executing: $($Invocation.MyCommand) $($params -join ' ')" -Level Verbose PositionMessage : At D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1:498 char:5 + Write-Log -Message "[$($Invocation.MyCommand.Module.Version)] Exe … + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ PSScriptRoot : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0 PSCommandPath : D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1 CommandOrigin : Internal ScriptStackTrace : at Write-InvocationLog, D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\Helpers.ps1: line 498 at Set-GitHubAuthentication, D:\home\data\ManagedDependencies\200908203829192.r\PowerShellForGitHub\0.14.0\GitHubConfiguration.ps1: line 877 at <ScriptBlock>, D:\home\site\wwwroot\Github\run.ps1: line 42

#### Suggested solution to the issue
None

#### Requested Assignment
    - Doesn't matter

#### Operating System
Key : OSVersion
Value : Microsoft Windows NT 10.0.14393.0
Name : OSVersion

Key : Is 64-bit
Value : True
Name : Is 64-bit

Key : Current culture
Value : en-US
Name : Current culture

Key : Current UI culture
Value : en-US
Name : Current UI culture

#### PowerShell Version
Key : PSVersion
Value : 7.0.2
Name : PSVersion

Key : PSEdition
Value : Core
Name : PSEdition

Key : GitCommitId
Value : 7.0.2
Name : GitCommitId

Key : OS
Value : Microsoft Windows 10.0.14393
Name : OS

Key : Platform
Value : Win32NT
Name : Platform

Key : PSCompatibleVersions
Value : {1.0, 2.0, 3.0, 4.0…}
Name : PSCompatibleVersions

Key : PSRemotingProtocolVersion
Value : 2.3
Name : PSRemotingProtocolVersion

Key : SerializationVersion
Value : 1.1.0.1
Name: SerializationVersion

Key : WSManStackVersion
Value : 3.0
Name : WSManStackVersion

#### Module Version
Running: Installed: 0.15.0

**Comment by HowardWolosky** [comment]
Hmmm... it looks like the issue is around logging, possibly with how it is trying to determine the path to the log file.
Can you provide the result of the following command (under the context of your azure function app)?:

```powershell
Get-GitHubConfiguration -Name LogPath
```

As a temporary measure to get unblocked, try running this first:.

```powershell
Set-GitHubConfiguration -DisableLogging
```

That should get you unblocked while I sort out what the right fix is.

Also -- can you provide any documentation/steps that you followed for creating your Azure Function App using PowerShellForGitHub so that I can attempt to reproduce your problem (so that I can validate any potential fix)?

**Comment by HowardWolosky** [comment]
Okay, I tried this out myself.
It appears that

```powershell
[System.Environment]::GetFolderPath('MyDocuments')
```

resolves to nothing in the Azure Function App.  That means that the default `logPath` ends up being empty:

https://github.com/microsoft/PowerShellForGitHub/blob/3094c9789caf00382134a0703344f655b13a1e33/GitHubConfiguration.ps1#L650-L657

https://github.com/microsoft/PowerShellForGitHub/blob/3094c9789caf00382134a0703344f655b13a1e33/GitHubConfiguration.ps1#L671

The potential options that I see for fixing this are:

1. Revert to using the temp folder when the documents folder is unavailable
   a. Just do that.
   b. Do this _and_ write a warning letting users know that they can change the path back to a non-temp location
2. Auto-disable logging and then inform users with a `Write-Warning` that logging has been disabled until they configure a valid `logPath` and then re-enable logging.

Option `1a` "just works" with the risk that the log file will be unexpectedly lost when temp files are cleared.  Option `1b` is mildly better because it lets users know what has happened, but in a non-interactive scenarios like an Azure Function App, users are unlikely to ever see the message.

Option `2` seems riskier because this is more likely to happen in non-interactive scenarios, and thus you may never realize that logging was disabled.  Option `1[a|b]` at least gives you the _possibility_ to still get access to the log after the fact.

I think I'll go with 1b.

**Comment by ferristocrat** [comment]
Thanks @HowardWolosky ! I was just finishing a reply to your first message.

I like your recommendation for 2a. For my use case 1a would probably suffice, and even be preferred, but 2b covers more ground

**Comment by HowardWolosky** [comment]
@ferristocrat :

> I like your recommendation for 2a. For my use case 1a would probably suffice, and even be preferred, but 2b covers more ground

I'm assuming that you were referring to the option `1`'s there, and not option `2`?

**Comment by ferristocrat** [comment]
> @ferristocrat :
> 
> > I like your recommendation for 2a. For my use case 1a would probably suffice, and even be preferred, but 2b covers more ground
> 
> I'm assuming that you were referring to the option `1`'s there, and not option `2`?

Yeah. Something like:
```powershell
if (-not [System.String]::IsNullOrEmpty($documentsFolder)) { 
     $logPath = Join-Path -Path $documentsFolder -ChildPath 'PowerShellForGitHub.log' 
} else {
     $logPath = Join-Path -Path $env:TEMP -ChildPath 'PowerShellForGitHub.log'
     Write-Warning "some warning"
}
```

**Comment by HowardWolosky** [comment]
The fix is a little more complicated than that, but yes.

The fix should get published by tomorrow.  In the meantime, it looks like the `DisableLogging` workaround won't work, because of this:

https://github.com/microsoft/PowerShellForGitHub/blob/3094c9789caf00382134a0703344f655b13a1e33/Helpers.ps1#L142

(which tries to assign an empty string in this scenario as a `FileInfo`).

Your workaround until the fix is in is to explicitly set the logPath yourself at the start of your script:

```powershell
Set-GitHubConfiguration -LogPath (Join-Path -Path $env:TEMP -ChildPath 'PowerShellForGitHub.log')
```

**Comment by ferristocrat** [comment]
Thanks so much!

**Comment by HowardWolosky** [comment]
[`0.15.1`](https://www.powershellgallery.com/packages/PowerShellForGitHub/0.15.1) has been published with the fix.  The workaround is no longer necessary.
